src/app/signup/page.tsx

```diff
@@ -1,22 +1,83 @@
 "use client";
 
-import Image from "next/image";
 import { GradButton } from "@/components/ui/grad-button";
 import SignUpImage from "@/components/ui/signup-image";
+import { signupUser } from "@/lib/User/actions";
+import {
+  EMAIL_ALREADY_EXISTS_MSG,
+  SIGNUP_MAX_EMAIL_LENGTH,
+  SIGNUP_MAX_NAME_LENGTH,
+  SIGNUP_MAX_PASSWORD_LENGTH,
+  SIGNUP_MIN_NAME_LENGTH,
+  SIGNUP_MIN_PASSWORD_LENGTH,
+  signUpSchema,
+} from "@/lib/User/validations";
 import { zodResolver } from "@hookform/resolvers/zod";
+import Image from "next/image";
+import { useMemo, useRef, useState } from "react";
+import { useFormState } from "react-dom";
 import { useForm } from "react-hook-form";
-import { signupUser } from "@/lib/User/actions";
-import { signUpSchema } from "@/lib/User/validations";
 
 const SignUp = () => {
+  const formRef = useRef<HTMLFormElement>(null);
+  const [state, formAction] = useFormState(signupUser, undefined);
+  const previousEmailValue = useMemo(() => {
+    return state
+      ? state.find((s) => s.field === "email")?.previousValue?.toString() || ""
+      : "";
+  }, [state]);
+  const previousFullNameValue = useMemo(() => {
+    return state
+      ? state.find((s) => s.field === "fullName")?.previousValue?.toString() ||
+          ""
+      : "";
+  }, [state]);
+  const previousPasswordValue = useMemo(() => {
+    return state
+      ? state.find((s) => s.field === "password")?.previousValue?.toString() ||
+          ""
+      : "";
+  }, [state]);
+  const previousConfirmPasswordValue = useMemo(() => {
+    return state
+      ? state
+          .find((s) => s.field === "confirmPassword")
+          ?.previousValue?.toString() || ""
+      : "";
+  }, [state]);
+  const emailError = useMemo(() => {
+    return state?.find((error) => error.field === "email")?.message;
+  }, [state]);
+  const fullNameError = useMemo(() => {
+    return state?.find((error) => error.field === "fullName")?.message;
+  }, [state]);
+  const passwordError = useMemo(() => {
+    return state?.find((error) => error.field === "password")?.message;
+  }, [state]);
+  const confirmPasswordError = useMemo(() => {
+    return state?.find((error) => error.field === "confirmPassword")?.message;
+  }, [state]);
+
+  const [showEmailStateError, setShowEmailStateError] = useState(
+    previousEmailValue !== "",
+  );
+
   const {
     register,
     handleSubmit,
     formState: { errors },
   } = useForm({
     resolver: zodResolver(signUpSchema),
+    defaultValues: {
+      fullName: previousFullNameValue || "",
+      email: previousEmailValue || "",
+      password: previousPasswordValue || "",
+      confirmPassword: previousConfirmPasswordValue || "",
+    },
   });
 
+  const emailValidator = register("email");
```

Comment:
So I guess relying on the default {...register()} usage, we can't do that here anymore?

Reply:
Probably. Since to keep track whether the email is the same as the previous state's, the onChange of the input needs to be accessed. We can't override it as well since it's being used by react-hook-form.